File: com.android.tools.build/4.0.1/com.android.tools/sdk-common/27.0.1/com/android/ide/common/gradle/model/IdeSourceProvider.java
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.ide.common.gradle.model;

import com.android.annotations.NonNull;
import com.android.builder.model.SourceProvider;
import com.google.common.collect.ImmutableList;
import java.io.File;
import java.io.Serializable;
import java.util.Collection;
import java.util.Collections;
import java.util.Objects;

/** Creates a deep copy of a {@link SourceProvider}. */
public final class IdeSourceProvider implements SourceProvider, Serializable {
    // Increase the value when adding/removing fields or when changing the serialization/deserialization mechanism.
    private static final long serialVersionUID = 3L;

    @NonNull private final String myName;
    @NonNull private final File myManifestFile;
    @NonNull private final Collection<File> myJavaDirectories;
    @NonNull private final Collection<File> myResourcesDirectories;
    @NonNull private final Collection<File> myAidlDirectories;
    @NonNull private final Collection<File> myRenderscriptDirectories;
    @NonNull private final Collection<File> myCDirectories;
    @NonNull private final Collection<File> myCppDirectories;
    @NonNull private final Collection<File> myResDirectories;
    @NonNull private final Collection<File> myAssetsDirectories;
    @NonNull private final Collection<File> myJniLibsDirectories;
    @NonNull private final Collection<File> myShadersDirectories;
    private final int myHashCode;

    // Used for serialization by the IDE.
    IdeSourceProvider() {
        myName = "";
        //noinspection ConstantConditions
        myManifestFile = null;
        myJavaDirectories = Collections.emptyList();
        myResourcesDirectories = Collections.emptyList();
        myAidlDirectories = Collections.emptyList();
        myRenderscriptDirectories = Collections.emptyList();
        myCDirectories = Collections.emptyList();
        myCppDirectories = Collections.emptyList();
        myResDirectories = Collections.emptyList();
        myAssetsDirectories = Collections.emptyList();
        myJniLibsDirectories = Collections.emptyList();
        myShadersDirectories = Collections.emptyList();

        myHashCode = 0;
    }

    public IdeSourceProvider(@NonNull SourceProvider provider) {
        myName = provider.getName();
        myManifestFile = provider.getManifestFile();
        myJavaDirectories = ImmutableList.copyOf(provider.getJavaDirectories());
        myResourcesDirectories = ImmutableList.copyOf(provider.getResourcesDirectories());
        myAidlDirectories = ImmutableList.copyOf(provider.getAidlDirectories());
        myRenderscriptDirectories = ImmutableList.copyOf(provider.getRenderscriptDirectories());
        myCDirectories = ImmutableList.copyOf(provider.getCDirectories());
        myCppDirectories = ImmutableList.copyOf(provider.getCppDirectories());
        myResDirectories = ImmutableList.copyOf(provider.getResDirectories());
        myAssetsDirectories = ImmutableList.copyOf(provider.getAssetsDirectories());
        myJniLibsDirectories = ImmutableList.copyOf(provider.getJniLibsDirectories());
        myShadersDirectories =
                IdeModel.copyNewProperty(
                        () -> ImmutableList.copyOf(provider.getShadersDirectories()),
                        Collections.emptyList());
        myHashCode = calculateHashCode();
    }

    @Override
    @NonNull
    public String getName() {
        return myName;
    }

    @Override
    @NonNull
    public File getManifestFile() {
        return myManifestFile;
    }

    @Override
    @NonNull
    public Collection<File> getJavaDirectories() {
        return myJavaDirectories;
    }

    @Override
    @NonNull
    public Collection<File> getResourcesDirectories() {
        return myResourcesDirectories;
    }

    @Override
    @NonNull
    public Collection<File> getAidlDirectories() {
        return myAidlDirectories;
    }

    @Override
    @NonNull
    public Collection<File> getRenderscriptDirectories() {
        return myRenderscriptDirectories;
    }

    @Override
    @NonNull
    public Collection<File> getCDirectories() {
        return myCDirectories;
    }

    @Override
    @NonNull
    public Collection<File> getCppDirectories() {
        return myCppDirectories;
    }

    @Override
    @NonNull
    public Collection<File> getResDirectories() {
        return myResDirectories;
    }

    @Override
    @NonNull
    public Collection<File> getAssetsDirectories() {
        return myAssetsDirectories;
    }

    @Override
    @NonNull
    public Collection<File> getJniLibsDirectories() {
        return myJniLibsDirectories;
    }

    @Override
    @NonNull
    public Collection<File> getShadersDirectories() {
        return myShadersDirectories;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (!(o instanceof IdeSourceProvider)) {
            return false;
        }
        IdeSourceProvider provider = (IdeSourceProvider) o;
        return Objects.equals(myName, provider.myName)
                && Objects.equals(myManifestFile, provider.myManifestFile)
                && Objects.equals(myJavaDirectories, provider.myJavaDirectories)
                && Objects.equals(myResourcesDirectories, provider.myResourcesDirectories)
                && Objects.equals(myAidlDirectories, provider.myAidlDirectories)
                && Objects.equals(myRenderscriptDirectories, provider.myRenderscriptDirectories)
                && Objects.equals(myCDirectories, provider.myCDirectories)
                && Objects.equals(myCppDirectories, provider.myCppDirectories)
                && Objects.equals(myResDirectories, provider.myResDirectories)
                && Objects.equals(myAssetsDirectories, provider.myAssetsDirectories)
                && Objects.equals(myJniLibsDirectories, provider.myJniLibsDirectories)
                && Objects.equals(myShadersDirectories, provider.myShadersDirectories);
    }

    @Override
    public int hashCode() {
        return myHashCode;
    }

    private int calculateHashCode() {
        return Objects.hash(
                myName,
                myManifestFile,
                myJavaDirectories,
                myResourcesDirectories,
                myAidlDirectories,
                myRenderscriptDirectories,
                myCDirectories,
                myCppDirectories,
                myResDirectories,
                myAssetsDirectories,
                myJniLibsDirectories,
                myShadersDirectories);
    }

    @Override
    public String toString() {
        return "IdeSourceProvider{"
                + "myName='"
                + myName
                + '\''
                + ", myManifestFile="
                + myManifestFile
                + ", myJavaDirectories="
                + myJavaDirectories
                + ", myResourcesDirectories="
                + myResourcesDirectories
                + ", myAidlDirectories="
                + myAidlDirectories
                + ", myRenderscriptDirectories="
                + myRenderscriptDirectories
                + ", myCDirectories="
                + myCDirectories
                + ", myCppDirectories="
                + myCppDirectories
                + ", myResDirectories="
                + myResDirectories
                + ", myAssetsDirectories="
                + myAssetsDirectories
                + ", myJniLibsDirectories="
                + myJniLibsDirectories
                + ", myShadersDirectories="
                + myShadersDirectories
                + '}';
    }
}

File: com.android.tools.build/4.2.0-alpha06/com.android.tools/sdk-common/27.2.0-alpha06/com/android/ide/common/gradle/model/IdeSourceProvider.kt
/*
 * Copyright (C) 2017 The Android Open Source Project
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.android.ide.common.gradle.model

import com.android.builder.model.SourceProvider
import com.google.common.annotations.VisibleForTesting
import java.io.File
import java.io.Serializable

/** Creates a deep copy of a [SourceProvider].  */
data class IdeSourceProvider(
  private val myName: String,
  private val myFolder: File?,
  private val myManifestFile: String,
  private val myJavaDirectories: Collection<String>,
  private val myResourcesDirectories: Collection<String>,
  private val myAidlDirectories: Collection<String>,
  private val myRenderscriptDirectories: Collection<String>,
  private val myCDirectories: Collection<String>,
  private val myCppDirectories: Collection<String>,
  private val myResDirectories: Collection<String>,
  private val myAssetsDirectories: Collection<String>,
  private val myJniLibsDirectories: Collection<String>,
  private val myShadersDirectories: Collection<String>,
  private val myMlModelsDirectories: Collection<String>
) : SourceProvider, Serializable {
  private fun String.translate(): File = (myFolder?.resolve(this) ?: File(this)).normalize()
  private fun Collection<String>.translate(): Collection<File> = map { it.translate() }

  companion object {
    @JvmStatic
    fun create(provider: SourceProvider, deduplicate: String.() -> String): IdeSourceProvider {
      val folder: File? = provider.manifestFile.parentFile

      fun File.makeRelativeAndDeduplicate(): String = (if (folder != null) relativeToOrSelf(folder) else this).path.deduplicate()
      fun Collection<File>.makeRelativeAndDeduplicate(): Collection<String> = map { it.makeRelativeAndDeduplicate() }

      return IdeSourceProvider(
        myName = provider.name,
        myFolder = folder,
        myManifestFile = provider.manifestFile.makeRelativeAndDeduplicate(),
        myJavaDirectories = provider.javaDirectories.makeRelativeAndDeduplicate(),
        myResourcesDirectories = provider.resourcesDirectories.makeRelativeAndDeduplicate(),
        myAidlDirectories = provider.aidlDirectories.makeRelativeAndDeduplicate(),
        myRenderscriptDirectories = provider.renderscriptDirectories.makeRelativeAndDeduplicate(),
        myCDirectories = provider.cDirectories.makeRelativeAndDeduplicate(),
        myCppDirectories = provider.cppDirectories.makeRelativeAndDeduplicate(),
        myResDirectories = provider.resDirectories.makeRelativeAndDeduplicate(),
        myAssetsDirectories = provider.assetsDirectories.makeRelativeAndDeduplicate(),
        myJniLibsDirectories = provider.jniLibsDirectories.makeRelativeAndDeduplicate(),
        myShadersDirectories = IdeModel.copyNewPropertyNonNull(
          { provider.shadersDirectories },
          emptyList()
        ).makeRelativeAndDeduplicate(),
        myMlModelsDirectories = IdeModel.copyNewPropertyNonNull(
          { provider.mlModelsDirectories },
          emptyList()
        ).makeRelativeAndDeduplicate()
      )
    }

    @JvmStatic
    @VisibleForTesting
    fun create(provider: SourceProvider): IdeSourceProvider =
      create(provider, deduplicate = { this })
  }

  // Used for serialization by the IDE.
  constructor() : this(
    myName = "",
    myFolder = File(""),
    myManifestFile = "",
    myJavaDirectories = mutableListOf(),
    myResourcesDirectories = mutableListOf(),
    myAidlDirectories = mutableListOf(),
    myRenderscriptDirectories = mutableListOf(),
    myCDirectories = mutableListOf(),
    myCppDirectories = mutableListOf(),
    myResDirectories = mutableListOf(),
    myAssetsDirectories = mutableListOf(),
    myJniLibsDirectories = mutableListOf(),
    myShadersDirectories = mutableListOf(),
    myMlModelsDirectories = mutableListOf()
  )

  override fun getName(): String = myName
  override fun getManifestFile(): File = myManifestFile.translate()
  override fun getJavaDirectories(): Collection<File> = myJavaDirectories.translate()
  override fun getResourcesDirectories(): Collection<File> = myResourcesDirectories.translate()
  override fun getAidlDirectories(): Collection<File> = myAidlDirectories.translate()
  override fun getRenderscriptDirectories(): Collection<File> = myRenderscriptDirectories.translate()
  override fun getCDirectories(): Collection<File> = myCDirectories.translate()
  override fun getCppDirectories(): Collection<File> = myCppDirectories.translate()
  override fun getResDirectories(): Collection<File> = myResDirectories.translate()
  override fun getAssetsDirectories(): Collection<File> = myAssetsDirectories.translate()
  override fun getJniLibsDirectories(): Collection<File> = myJniLibsDirectories.translate()
  override fun getShadersDirectories(): Collection<File> = myShadersDirectories.translate()
  override fun getMlModelsDirectories(): Collection<File> = myMlModelsDirectories.translate()
}